flfm/app/main.py

```diff
@@ -0,0 +1,470 @@
+    Input(dict(type="image-slider", index=MATCH), "value"),
+    State(dict(type="image-slider", index=MATCH), "tooltip"),
+    State(dict(type="n-frames", index="psf"), "data"),  # NOTE: This intentionally uses index="psf".
+    State("depth-step-size", "value"),
+    State(dict(type="color-scale", index=MATCH), "value"),
+    State(dict(type="image-slider", index=MATCH), "id"),
+    prevent_initial_call=True,
+)
+def update_image_from_slider(value, tooltip, n_frames, step_size, color_scale, id):
+    """Re-plot image based on frame/depth selected from slider."""
+    global image_data
+    # tooltip.transform (clientside js )would render faster than updating the template in a server side callback,
+    # however, it only takes a single input ``value`` and there doesn't seem a trivial way to also pass ``step_size``
+    # and ``n_frames`` to the js transform function (and have it dynamically served/reloaded).
+    tooltip["template"] = slider_template_str(value, _center_frame(n_frames), step_size)
+    fig = plot_image(image_data[id["index"]][value, :, :], color_scale=color_scale)
+    return fig, tooltip
+
+
+@callback(
+    Output({"type": "image-slider", "index": ALL}, "tooltip", allow_duplicate=True),
+    Input("depth-step-size", "value"),
+    State({"type": "image-slider", "index": ALL}, "value"),
+    State({"type": "image-slider", "index": ALL}, "tooltip"),
+    State({"type": "n-frames", "index": "psf"}, "data"),
+    prevent_initial_call=True,
+)
+def update_slider_tooltip(depth_step_size, values, tooltips, n_frames):
+    for i, value in enumerate(values):
+        tooltips[i]["template"] = slider_template_str(value, _center_frame(n_frames), depth_step_size)
+    return tooltips
+
+
+def _display_image_with_slider(data, id):
+    """Plot image and display frame/depth slider (where appropriate)."""
+    if data is None:
+        return no_update
+
+    n_frames = len(data)
+    center_frame = _center_frame(n_frames)
+    default_slider_value = center_frame
+    fig = plot_image(data[center_frame, :, :])
+    children = [
+        dcc.Graph(id=dict(type="image-graph", index=id), figure=fig, responsive=True, style={"height": "100vh"}),
+        html.Label("Select Image Frame/Depth"),
+        dcc.Slider(
+            0,
+            len(data) - 1,
+            step=1,
+            value=default_slider_value,
+            id=dict(type="image-slider", index=id),
+            disabled=n_frames <= 1,
+            tooltip={
+                "always_visible": n_frames > 1,
+                "template": slider_template_str(default_slider_value, center_frame, app_settings.DEPTH_STEP_SIZE),
+            },
+        ),
+        html.Label("Color Scale"),
+        dcc.Dropdown(
+            id=dict(type="color-scale", index=id),
+            options=colorscales,
+            value=app_settings.IMSHOW_COLOR_SCALE,
+        ),
+    ]
+    return children, n_frames
+
+
+def upload_file(contents, filename):
```

Comment:
Can you add some type hints here if the types are consistent and reasonable to reference?